text/0000-const-static-type-elision.md

```diff
@@ -0,0 +1,153 @@
+type, including in traits/impls.
+
+# Motivation
+[motivation]: #motivation
+
+In most cases, the type for constant and static items is obvious, and requiring
+a redundant type annotation is a papercut many programmers would want to avoid.
+For example, these two declarations would result in compiler errors in the
+current version of Rust:
+
+```rust
+const GREETING = "Hello, world!"; // unique type: &'static str
+static NUM = 42i32; // unique type: i32
+```
+
+This is usually no more than a small annoyance, but the risk involved in eliding
+the types also seems small.
+
+In the terms of
+[the ergonomics initiative blog post](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html),
+this change is broadly applicable, but the power is restrained by the
+limitations on type inference described below.
+
+# Detailed design
+[design]: #detailed-design
+
+Rust would allow `const` and `static` items to elide type annotations and infer
+the type, but only if type inference can infer a unique type for the expression
+*before* applying any fallback rules. So if we have the following items:
+
+```rust
+struct S {
+  a: i32
+}
+
+const A: bool = true;
+const B: i32 = 42i32;
+const C: &str = "hello";
+const D: S = S { a: 1 };
+const E: [f32; 2] = [1.0f32, 2.5f32];
+```
+
+They could be written like this:
+```rust
+struct S {
+  a: i32
+}
+
+const A = true;
+const B = 42i32;
+const C = "hello";
+const D = S { a: 1 };
+const E = [1.0f32, 2.5f32];
+```
+
+To minimize the reasoning footprint, type elision would use only local type
+inference, rather than attempting to infer a type based on a later use of the
+item as with `let`-bound variables. For example, the following would result in a
+type error, because there are multiple possible types for the literal 42
+(e.g. `i16`, `i32`, etc.), even though the use in `get_big_number` would require
+it to be `i64`.
+
+```rust
+const THE_ANSWER = 42; // nothing in RHS indicates this must be i16
+
+fn get_big_number() -> i16 {
+    THE_ANSWER
+}
+```
+
+## Integer/Float Fallback
+
+The fallback rules (specifically, defaulting integer literals to `i32` and float
+literals to `f64`) are disallowed in cases where multiple typings are valid to
+prevent the type of an exported item from changing only by removing a type
+annotation. For example, say some crate exports the following:
+
+```rust
+const X: i64 = 5;
+```
+
+If the developer later decides to elide the type annotation, then fallback would
+infer the type of `X` as `i32` rather than `i64`. If `X` is exported but not
+used within the crate, then this change could break downstream code without the
+crate developer realizing it. Admittedly, that scenario is unlikely, but
+ruling out fallback is the most conservative option and could always be added
+back in later.
+
+Fallback is acceptable, however, if the overall type is still unique even
+without the fallback rules, as in this example:
```

Comment:
Your assumptions are correct: `A` should type-check, but `B` should not.

According to my understanding of Rust's type inference mechanisms, this is always enforceable. Either the unification algorithm comes up with a unique type, in which case it's okay, or it doesn't, in which case the expression results in a type error. My understanding is limited, though, so I'd be happy to learn about any counterexamples.